### ISSUE TITLE
converting matconvnet models to caffe

### ISSUE THREAD

**Issue by jboys** [vlfeat/matconvnet#239]
My issue/question is the inverse of [issue #184](https://github.com/vlfeat/matconvnet/issues/184):

I have trained a model using matconvnet that I would like to use within caffe.

I understand that it is possible to import caffe models for use in matconvnet using the [import-caffe.py](https://github.com/vlfeat/matconvnet/blob/master/utils/import-caffe.py) script.

Does there exist a way to do the reverse and export a matconvnet model for use within caffe?

**Comment by vedaldi** [comment]
Hi, currently we have no script for that, but it should be relatively easy to do, particularly using Caffe-MATLAB bindings.

**Comment by vidyagayash** [comment]
Hi,

I have been using MatConvNet successfully. I am trying to replicate the learning in Caffe. However I am unable to find out what the exact learning parameters were used in MatConvNet with respect to Caffe parameters. Like what are the default values for learning policy type (fixed, inv, step etc) and the step size, gamma, power etc, for MatConvNet ? Where do I find it ?  

Thanks
CleverCrow

**Comment by vedaldi** [comment]
Hi, check out the cnn_train.m script in examples.

You can specify a learning schedule, adjusting the learning rate after each iteration (this allows to implement any fixed schedule in practice). Momentum is also defined there.

**Comment by zoharby** [comment]
see PR #298, I wrote a function to export the models to Caffe:
https://github.com/zoharby/matconvnet/blob/master/utils/convert_matconvnet_caffe.m

**Comment by hyojinie** [comment]
@zoharby I keep notice differences in result even after weight transfers are successful. It works, but assertion (assert(max_diff <= 1e-5)) keep fails. Sometimes I observe values differ by 50~100 after one convolution layer...

**Comment by zoharby** [comment]
@hyojinie - probably my script is incompatible with something in you network architecture.
Can you send me a link (Dropbox/Google Drive/etc.) to download a mat file of the network you're trying to convert and a sample input? I will have a look.

**Comment by hyojinie** [comment]
@zoharby Oh I see. I was running in the CPU mode. So I guess there is a bug in MatConvNet in cpu mode. Thanks for the comment!

**Comment by siebenkopf** [comment]
@zoharby Thank you for the code to export a network to Caffe, this was what I was searching for. Unfortunately, the function only accepts SimpleNN's, but I have a DagNN, so I cannot use your function. Do you think it would be possible to provide a similar function to convert a DagNN to Caffe? I mean, I could write one myself, I guess, but my Matlab knowledge is limited...

Or is there another way to convert a DagNN into a SimpleNN? I have seen that there are conversion functions for the other way around, but these don't help me...

Thanks in advance
Manuel

**Comment by zoharby** [comment]
@siebenkopf Currently there is no support for DagNNs. I've just started playing around with DagNNs myself, so might end up adapting the export script to support DagNNs as well. 

Note that I'm not sure how much of it is supported by caffe, so a generic script might be a hard task...

**Comment by Rizlene** [comment]
Hi,

I'm new in Deep Learning and I want to convert a network from MatConvNet to Caffe. I try to use your script but it doesn't work and don't know what I'm doing wrong?

The network i'm trying to convert is the one taken from the paper "Deep Features for Text Spotting, Jaderberg, M. and Vedaldi, A. and Zisserman, A."
And i have this error : 
Attempt to reference field of non-structure array.

Error in export_matconvnet_caffe (line 34)
for idx = 1:length(net.layers)

Thanks in advance!!
R.O

**Comment by Piyush3dB** [comment]
@Rizlene have you managed to get it working in matconvnet first?

**Comment by Rizlene** [comment]
Yes, the network works well in matconvnet!

Thks

**Comment by zoharby** [comment]
So you're probably using the export function incorrectly. The network ("net") should have a field called "layers". It seems you are passing something else to the function instead of the MatConvNet network.

**Comment by Rizlene** [comment]
Ok, I was sure that it was the problem and i'm not sure that it is possible to solve it easily.

as I said, I work with the network described on paper  "Deep Features for Text Spotting, Jaderberg, M. and Vedaldi, A. and Zisserman, A., eccv2014" and the only thing I have is a file ".mat" which are described each layer (layer0, layer1 ... layer19). But actually there is no field called "layers".

You can find the network here : https://bitbucket.org/jaderberg/eccv2014_textspotting/src/581855bba39e662d66b05392b864dfbdd15d4142/models/?at=master
I tried to convert the one called "detnet_layers.mat", but they all have substantially the same structure

**Comment by zoharby** [comment]
This looks like some ancient version of MatConvNet, which doesn't even resemble the current structure...
If you can get it to work with the current version of MatConvNet then doing the conversion should be easy from there. You'll probably have to build it layer by layer from the data you have in the mat files.

**Comment by Rizlene** [comment]
Ok! Thank you. I'll try to do this ;)

**Comment by azanbar** [comment]
Hi, 
I am trying to convert a matconvnet type network to Caffe, which uses the layer bnorm. 

Does anyone know how to convert this layer to Caffe? 
I saw an example of using BatchNorm in Caffe, but it doesn't mention about fields like weights and momentum:
 
`layer {
  name: "bn1"
  type: "BatchNorm"
  bottom: "pool1"
  top: "bn1"
  param {
    lr_mult: 0
  }
  param {
    lr_mult: 0
  }
  param {
    lr_mult: 0
  }
}`

In addition, couldn't find this option in the script that @zoharby wrote.

Thank you very much

**Comment by lenck** [comment]
Hi, if you don't need to train it in caffe, the simplest would be to adjust your network with [cnn_imagenet_deploy](https://github.com/vlfeat/matconvnet/blob/master/examples/imagenet/cnn_imagenet_deploy.m) as it removes the batch normalisation and puts its parameters into the closest convolution...

**Comment by azanbar** [comment]
thanks a lot @lenck, 
is it ok to merge the weight of the bnorm layer with the existed weights in the previous conv layer?

**Comment by azanbar** [comment]
hi @lenck ,
thank you again for your help. 
at first I couldn't get the same results from the original network (with the bnorm layers) and the merged network, but it was fixed after setting the network on 'test' mode

**Comment by ainilian** [comment]
hi @Rizlene ，
i just stepped into deep learning. i am trying to convert the eccv2014-textspotting models to caffe . i have same problem as you . 
if you done it . can you give me some indication?

**Comment by Rizlene** [comment]
hi @ainilian ,

I'm sorry i can't help you. I've tried to update the network, to match it with current structure of MatConvNet without success :(. I also tried to convert it manually without using the script of zoharby, but I had problems to build the .caffemodel file.

So I gave up for the moment. If you have any other ideas, i'm always interested.

Best of luck!!!

Rizlene

**Comment by ainilian** [comment]
Thank you all the same !!! @Rizlene

**Comment by lenck** [comment]
Hi, the conversion script for simplenn has been merged to the devel branch... (see [simplenn_caffe_deploy](https://github.com/vlfeat/matconvnet/blob/devel/utils/simplenn_caffe_deploy.m) )

**Comment by azanbar** [comment]
hi, 
i am trying to converted trained MatConvNet DagNN to Caffe - does any one have an example of such net in Caffe?
is it possible?

thanks a lot..

**Comment by KranthiGV** [comment]
@Rizlene @ainilian 
Has anyone of you successfully converted it?
Also, can you let me know if you are able to run the [Synthetic Data and Artificial Neural Networks for Natural Scene Text Recognition model](http://www.robots.ox.ac.uk/~vgg/research/text/#sec-models) using the MatConvNet provided in the package?
Thank you!

**Comment by lizhengwei1992** [comment]
@siebenkopf  @zoharby 
I have the same problem about converting DagNN MatConvNet  to caffemodel, do you slove the problem now ? 
Thanks in advance!

**Comment by llmpass** [comment]
Hi @zoharby, I'm new to matconvnet and I'm using your script to convert .mat model to caffe model. I got this problem:  
error: structure has no member 'layers'
error: called from
    convert_matconvnet_caffe at line 35 column 1

any ideas?

**Comment by zoharby** [comment]
@llmpass - please use the conversion script which is now part of the master branch:
https://github.com/vlfeat/matconvnet/blob/devel/utils/simplenn_caffe_deploy.m
It should hopefully be a bit more robust.

**Comment by llmpass** [comment]
@zoharby Thank you very much for your quick response!

**Comment by Shoshin23** [comment]
Hi @Rizlene I'm looking for caffe versions of the exact same .mat file. Any luck converting yours? Would you be able to upload the converted model someplace?

**Comment by BigJerry** [comment]
anyone fix the problem about converting bnorm layer from matconvnet model to caffe model?
